quickwit-core/src/index.rs

```diff
@@ -65,28 +69,114 @@ pub async fn search_index(metastore_uri: &str, index_id: &str) -> anyhow::Result
 ///
 /// * `metastore_uri` - The metastore Uri for accessing the metastore.
 /// * `index_id` - The target index Id.
+/// * `dry_run` - Should this only return a list of affected files without performing deletion.
 ///
-pub async fn delete_index(metastore_uri: &str, index_id: &str) -> anyhow::Result<()> {
+pub async fn delete_index(
+    metastore_uri: &str,
+    index_id: &str,
+    dry_run: bool,
+) -> anyhow::Result<Vec<PathBuf>> {
     let metastore = MetastoreUriResolver::default()
         .resolve(&metastore_uri)
         .await?;
+    let storage_resolver = Arc::new(StorageUriResolver::default());
+
+    if dry_run {
+        return list_index_files(metastore_uri, index_id, storage_resolver).await;
+    }
+
+    // schedule all staged & published splits for delete
+    let mut active_splits = metastore
+        .list_splits(index_id, SplitState::Published, None)
+        .await?;
+    let staged_splits = metastore
+        .list_splits(index_id, SplitState::Staged, None)
+        .await?;
+    active_splits.extend(staged_splits);
+
+    let split_ids = active_splits
+        .iter()
+        .map(|split_meta| split_meta.split_id.as_str())
+        .collect::<Vec<_>>();
+    metastore
+        .mark_splits_as_deleted(index_id, split_ids)
+        .await?;
+
+    let files = garbage_collect(metastore_uri, index_id, storage_resolver).await?;
     metastore.delete_index(index_id).await?;
```

Comment:
We still have a race condition here. If someone adds a new split right before we delete the index, we will end up with a orphan split in the storage.

Reply:
Can you add a TODO here to fix that later?
It is not a major bug in the sense, user are a bit at fault here.

Reply:
Do you mean when we a user runs the `index` command on another terminal while the `delete` was being performed?

Reply:
yes exactly

Reply:
or another user